File: csharp/allergies/Allergies.cs
using System;
using System.Linq;

[Flags]
public enum Allergen
{
    Eggs = 1,
    Peanuts = 2,
    Shellfish = 4,
    Strawberries = 8,
    Tomatoes = 16,
    Chocolate = 32,
    Pollen = 64,
    Cats = 128
}

public class Allergies
{
    private Allergen _allergen;

    public Allergies(int mask)
    {
        _allergen = (Allergen)mask;
    }

    public bool IsAllergicTo(Allergen allergen)
    {
        return _allergen.HasFlag(allergen);
    }

    public Allergen[] List()
    {
        // Could also precalculate array in constructor and then return copy of the array.
        return Enum.GetValues(typeof(Allergen))
            .Cast<Allergen>()
            .Where(e => _allergen.HasFlag(e))
            .ToArray();
    }
}

File: zig/allergies/allergies.zig
const std = @import("std");
const EnumSet = std.EnumSet;

pub const Allergen = enum(u8) {
    eggs = 1,
    peanuts = 2,
    shellfish = 4,
    strawberries = 8,
    tomatoes = 16,
    chocolate = 32,
    pollen = 64,
    cats = 128,
};

pub fn isAllergicTo(score: u8, allergen: Allergen) bool {
    const allergen_set = initAllergenSet(score);
    return allergen_set.contains(allergen);
}

pub fn initAllergenSet(score: usize) EnumSet(Allergen) {
    var allergen_set = EnumSet(Allergen).initEmpty();

    inline for (std.meta.fields(Allergen)) |allergen| {
        if (is_set(score, allergen.value)) {
            allergen_set.insert(@enumFromInt(allergen.value));
        }
    }

    return allergen_set;
}

inline fn is_set(value: usize, bit_mask: usize) bool {
    return value & bit_mask != 0;
}
